Fix resurrection logic to trigger early enough

albanD opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #137267

Fixes https://github.com/pytorch/pytorch/issues/136358

The bug here is that the Tensor object is actually 2 classes: `Tensor` from `_tensor.py` and `TensorBase` from c++.

Before this PR, they have the following gc methods:
Tensor:
 - tp_clear subtype_clear
 - tp_traverse THPVariable_subclass_traverse
 - tp_dealloc THPVariable_subclass_dealloc

TensorBase:
- tp_clear THPVariable_clear
- tp_traverse THPFunction_traverse (fake function that is just an error)
- tp_dealloc object_dealloc

The problem is that when clear is called on the Tensor, subtype_clear is going to clear the things owned by the "Tensor" type, in particular, its `__dict__` attribute, before delegating to the TensorBase clear where we detect that resurrection needs to happen and skip it.

albanD committed:
Fix resurrection logic to trigger early enough

[ghstack-poisoned]

kshitij12345 commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -2277,9 +2283,17 @@ int THPVariableMetaType_init(PyObject* cls, PyObject* args, PyObject* kwargs) {
   ((PyTypeObject*)cls)->tp_dealloc = (destructor)THPVariable_subclass_dealloc;
   ((PyTypeObject*)cls)->tp_traverse =
       (traverseproc)THPVariable_subclass_traverse;
+  ((PyTypeObject*)cls)->tp_clear = (inquiry)THPVariable_clear;
```
I was wondering if this should be renamed to `THPVariable_subclass_clear` for consistency.

albanD committed:
Update on "Fix resurrection logic to trigger early enough"


Fixes https://github.com/pytorch/pytorch/issues/136358

The bug here is that the Tensor object is actually 2 classes: `Tensor` from `_tensor.py` and `TensorBase` from c++.

Before this PR, they have the following gc methods:
Tensor:
 - tp_clear subtype_clear
 - tp_traverse THPVariable_subclass_traverse
 - tp_dealloc THPVariable_subclass_dealloc

TensorBase:
- tp_clear THPVariable_clear
- tp_traverse THPFunction_traverse (fake function that is just an error)
- tp_dealloc object_dealloc

The problem is that when clear is called on the Tensor, subtype_clear is going to clear the things owned by the "Tensor" type, in particular, its `__dict__` attribute, before delegating to the TensorBase clear where we detect that resurrection needs to happen and skip it.

[ghstack-poisoned]

ezyang reviewed (approved):
whoops! Glad it was simple

kshitij12345 reviewed (approved):
Thank you very much for the quick fix @albanD

albanD committed:
Update on "Fix resurrection logic to trigger early enough"


Fixes https://github.com/pytorch/pytorch/issues/136358

The bug here is that the Tensor object is actually 2 classes: `Tensor` from `_tensor.py` and `TensorBase` from c++.

Before this PR, they have the following gc methods:
Tensor:
 - tp_clear subtype_clear
 - tp_traverse THPVariable_subclass_traverse
 - tp_dealloc THPVariable_subclass_dealloc

TensorBase:
- tp_clear THPVariable_clear
- tp_traverse THPFunction_traverse (fake function that is just an error)
- tp_dealloc object_dealloc

The problem is that when clear is called on the Tensor, subtype_clear is going to clear the things owned by the "Tensor" type, in particular, its `__dict__` attribute, before delegating to the TensorBase clear where we detect that resurrection needs to happen and skip it.

[ghstack-poisoned]

albanD commented:
@pytorchbot merge -r

albanD committed:
Update

[ghstack-poisoned]

kit1980 commented:
2.5.1 is an emergency patch release to address known large regressions, moving this to 2.6.0

ezyang commented:
@kit1980 if you're gonna move this fix to 2.6.0, you might as well say there's only one patch going in 2.5.1 lol. This one is clear, unambiguous and solves silent corruption (that happens rarely, but it is silent corruption!)

kit1980 commented:
@ezyang We were considering actually to do only one patch, but decided to include also couple of fixes for critical issues where external users loudly complained after 2.5.0.
We want to release 2.5.1 next week, and creating new release candidate will delay it and increase risk.
We might do 2.5.2 later.

kit1980 commented:
Also, is this a regression introduced in 2.5.0?
If not, then it's not for a patch release.
